### PR TITLE
[FakeTensor]: Support fast_bind used in normalize_function

### PR DESCRIPTION
**Pull request by armijo1122**
## Motivation                                                                                                                                      
                                                                                                                                                     
  torch.fx.operator_schemas uses inspect.Signature.bind() in hot paths to (1) identify a matching overload/schema and (2) normalize args/kwargs into 
  a canonical form. While Signature.bind() is correct and general, it is relatively expensive for the limited signature shapes we typically handle   
  here.                                                                                                                                              
                                                                                                                                                     
  This overhead is especially visible in FakeTensor-heavy workflows (e.g., FakeTensor dispatch, meta/fake execution, and tracing paths), where       
  overall runtime is dominated by Python-side overhead rather than compute. Repeated signature binding can become a measurable contributor to end-to-
  end time.     

<img width="3070" height="818" alt="image" src="https://github.com/user-attachments/assets/b25c24fb-0659-4cc2-a19a-b87952492302" />
                                                                                                                                     
                                                                                                                                                     
  ## What this PR does                                                                                                                               
                                                                                                                                                     
  • Introduces _fast_bind(sig, args, kwargs) in torch/fx/operator_schemas.py as a fast-path subset of inspect.Signature.bind():                      
    • Supports POSITIONAL_ONLY, POSITIONAL_OR_KEYWORD, and KEYWORD_ONLY parameters (including defaults and common error cases).                      
    • Falls back to sig.bind(*args, **kwargs) for signatures containing *args / **kwargs (VAR_POSITIONAL / VAR_KEYWORD) to preserve correctness.     
  • Replaces multiple sig.bind(...) call sites in schema matching and normalization logic with _fast_bind(...).                                      
  • Adds a dedicated unit test suite to validate behavior matches Signature.bind() across a broad set of scenarios, including:                       
    • positional-only parameters (including the common "from"-style schema pattern),                                                                 
    • keyword-only parameters,                                                                                                                       
    • defaults / missing args / conflicting args,                                                                                                    
    • fallback behavior for VAR_POSITIONAL / VAR_KEYWORD.                                                                                            
                                                                                                                                                     
                                                                                                                                                     
  ## Why this is safe                                                                                                                                
                                                                                                                                                     
  • The fast path targets the parameter kinds already expected/handled by operator_schemas.py.                                                       
  • Complex signatures explicitly fall back to Signature.bind() to avoid semantic drift.                                                             
  • Tests compare BoundArguments results (including apply_defaults() behavior) against Signature.bind().                                             
                                                                                                                                                     
  ## Expected impact                                                                                                                                 
                                                                                                                                                     
  • Reduced Python overhead in FX operator schema matching and argument normalization(20x faster).                                                               
  • Improved throughput for FakeTensor-heavy execution (FakeTensor dispatch/tracing), where repeated binding costs are amplified.
  • End-to-End duration 5x faster for 10B model
<img width="2722" height="1168" alt="image" src="https://github.com/user-attachments/assets/a49d6c92-c306-4767-b3ec-0fe178a2b9e0" />

@bobrenjc93 @laithsakka could you help review this PR, thanks!

### CODE REVIEW

File: torch/fx/operator_schemas.py [hunk 1]
```diff
@@ -42,6 +42,86 @@ class ArgsKwargsPair(NamedTuple):
 _manual_overrides: dict[Callable, list[inspect.Signature]] = {}
 
 
+def _fast_bind(
```
Comment (laithsakka) [inline torch/fx/operator_schemas.py]: i see bind is used so much even outside schema is this something that would work only with operator scheme or general that we can use it to replace any call to bind

Comment (armijo1122) [reply torch/fx/operator_schemas.py]: @laithsakka Actually, the bind implementation itself is general-purpose and can be used to replace any bind call. However, the current bottleneck we are facing is specifically related to fake tensor overhead. To minimize risk and scope, I've only replaced the bind calls within the operator schema for now.

Comment (armijo1122) [reply torch/fx/operator_schemas.py]: @laithsakka @zou3519 @aorenste Could you please help review it when you have time? Thanks again for your help!

Comment (aorenste) [inline torch/fx/operator_schemas.py]: This is a bunch of code which isn't directly related to operator_schemas - let's move it to its own utility file.

Also consider having the signature match `.bind` so it's a more obvious replacement.

File: torch/fx/operator_schemas.py [hunk 2]
```diff
@@ -42,6 +42,86 @@ class ArgsKwargsPair(NamedTuple):
...
+    sig: inspect.Signature, args: tuple[Any, ...], kwargs: dict[str, Any]
+) -> inspect.BoundArguments:
+    """`inspect.Signature.bind` fast path
```
Comment (aorenste) [inline torch/fx/operator_schemas.py]: needs a better doc. Claude suggests:
```
      """Fast path for inspect.Signature.bind() for signatures without
      VAR_POSITIONAL or VAR_KEYWORD parameters. Falls back to sig.bind()
      for signatures that contain *args or **kwargs.
      """
```

File: test/fx/test_operator_schemas_fast_bind.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
+# Owner(s): ["module: fx"]
+
+import inspect
+import unittest
+
+import torch.fx.operator_schemas as op_schemas
+
+
+class TestFastBind(unittest.TestCase):
```
Comment (aorenste) [inline test/fx/test_operator_schemas_fast_bind.py]: Should probably be `torch.testing._internal.common_utils.TestCase` to be consistent with other tests in this dir.

File: torch/fx/operator_schemas.py [hunk 4]
```diff
@@ -42,6 +42,86 @@ class ArgsKwargsPair(NamedTuple):
...
+    len_args = len(args)
+
+    # max number of positional arguments (pos-only + pos-or-kw)
+    max_positional = 0
+    for p in params:
+        if p.kind in (
+            inspect.Parameter.POSITIONAL_ONLY,
+            inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        ):
+            max_positional += 1
```
Comment (aorenste) [inline torch/fx/operator_schemas.py]: Pull this out into a cached helper function. Something like:

```
# at the top-level
@functools.cache
def _signature_metadata(sig: inspect.Signature) -> tuple[tuple[dict[str, Parameter]], bool, int]:
    (returns tuple(sig.parameters.values()) and if any has VAR_POSITIONAL or VAR_KEYWORD, and the max_positional)
```
    Then the rest of this can just check the return of that...

**Review by aorenste — CHANGES_REQUESTED** [review]
Is there a test for `f()`? I may have missed it.

File: torch/fx/operator_schemas.py (hunk 1 above)
Comment (armijo1122) [reply torch/fx/operator_schemas.py]: just modified the signature to match .bind

File: torch/fx/operator_schemas.py (hunk 2 above)
Comment (armijo1122) [reply torch/fx/operator_schemas.py]: Thanks for your better doc, I applied this doc in newest version

File: test/fx/test_operator_schemas_fast_bind.py (hunk 3 above)
Comment (armijo1122) [reply test/fx/test_operator_schemas_fast_bind.py]: has been fixed in newest version

File: torch/fx/operator_schemas.py (hunk 4 above)
Comment (armijo1122) [reply torch/fx/operator_schemas.py]: I've used an lru cache in newest version, thanks for your great advice

File: torch/fx/operator_schemas.py (hunk 1 above)
Comment (armijo1122) [reply torch/fx/operator_schemas.py]: I've moved _fast_bind to torch/utils and move test case to test/ dir, thanks for your review!

File: torch/utils/_inspect.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+import inspect
+import functools
+from typing import Any
+
+
+@functools.lru_cache(maxsize=None)
+def _signature_metadata(
+    sig: inspect.Signature,
+) -> tuple[list[inspect.Parameter], bool, int]:
```
Comment (aorenste) [inline torch/utils/_inspect.py]: Should return tuple so it's safe to cache (list can be mutated, tuple can't). Also that's what the doc claims it returns.

```suggestion
) -> tuple[tuple[inspect.Parameter], bool, int]:
```

File: torch/utils/_inspect.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Returns tuple(sig.parameters.values()), if any has VAR_POSITIONAL or VAR_KEYWORD, and the max_positional
+    """
+    params = list(sig.parameters.values())
```
Comment (aorenste) [inline torch/utils/_inspect.py]:
```suggestion
    params = tuple(sig.parameters.values())
```

File: test/test_inspect_utils.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                _fast_bind(sig, *args, **kwargs)
+            return
+
+        got = _fast_bind(sig, *args, **kwargs)
+        self.assertEqual(ref.arguments, got.arguments)
+        self.assertEqual(ref.args, got.args)
+        self.assertEqual(ref.kwargs, got.kwargs)
+
+        # Also validate default population matches
+        ref = sig.bind(*args, **kwargs)
```
Comment (aorenste) [inline test/test_inspect_utils.py]: nit: Why do you re-call `sig.bind()` here instead of just reusing the existing `ref`? (Same Q for `got` below)

File: test/test_inspect_utils.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        ref.apply_defaults()
+        got = _fast_bind(sig, *args, **kwargs)
+        got.apply_defaults()
+        self.assertEqual(ref.arguments, got.arguments)
+        self.assertEqual(ref.args, got.args)
+        self.assertEqual(ref.kwargs, got.kwargs)
+
+    def test_positional_or_keyword_and_defaults(self):
```
Comment (aorenste) [inline test/test_inspect_utils.py:31]: I'm not seeing a test for an empty function. Something like:
```
	def test_zero_params(self):
	     def f():
	         pass
	     sig = inspect.signature(f)
	     self._assert_fast_bind_matches_sig_bind(sig, (), {})
	     self._assert_fast_bind_matches_sig_bind(sig, (1,), {})  # too many args
	     self._assert_fast_bind_matches_sig_bind(sig, (), {"x": 1})  # unexpected kwarg
```

File: test/test_inspect_utils.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
+# Owner(s): ["module: fx"]
+
+import inspect
+import unittest
+
+from torch.utils._inspect import _fast_bind
+from torch.testing._internal.common_utils import TestCase
```
Comment (aorenste) [inline test/test_inspect_utils.py]:
```suggestion
from torch.testing._internal.common_utils import TestCase, run_tests
```

File: test/test_inspect_utils.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            pass
+
+        sig = inspect.signature(f)
+        # Extra positional should be captured by *args; extra keywords by **kwargs
+        self._assert_fast_bind_matches_sig_bind(sig, (1, 2, 3), {"b": 4, "x": 5})
+        self._assert_fast_bind_matches_sig_bind(sig, (), {"a": 1, "x": 2})
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (aorenste) [inline test/test_inspect_utils.py]:
```suggestion
    run_tests()
```

**Review by aorenste — CHANGES_REQUESTED** [review]
Looking pretty good. A few small additional changes.

File: torch/utils/_inspect.py (hunk 5 above)
Comment (armijo1122) [reply torch/utils/_inspect.py]: Great catch! I totally missed this before

Comment (armijo1122) [reply torch/utils/_inspect.py]: I fixed it in the newest version

File: torch/utils/_inspect.py (hunk 6 above)
Comment (armijo1122) [reply torch/utils/_inspect.py]: I've fixed it in newest version, thanks for your suggestion!

File: test/test_inspect_utils.py (hunk 7 above)
Comment (armijo1122) [reply test/test_inspect_utils.py]: you're right, it's a duplicated call. I've fixed it

Comment (armijo1122) [reply test/test_inspect_utils.py]: so did the `got` below

File: test/test_inspect_utils.py (hunk 8 above)
Comment (armijo1122) [reply test/test_inspect_utils.py:31]: I have added the following test to cover zero-param scenario:

```python
   def test_zero_params(self):
        def f():
            pass

        sig = inspect.signature(f)
        self._assert_fast_bind_matches_sig_bind(sig, (), {})
        self._assert_fast_bind_matches_sig_bind(sig, (1,), {})  # too many args
        self._assert_fast_bind_matches_sig_bind(sig, (), {"x": 1})  # unexpected kwarg
        self._assert_fast_bind_matches_sig_bind(
            sig, (1,), {"b": 1}
        )  # unexpected arg & kwarg
```

File: test/test_inspect_utils.py (hunk 9 above)
Comment (armijo1122) [reply test/test_inspect_utils.py]: Fixed! Thanks again for catching this.

File: test/test_inspect_utils.py (hunk 10 above)
Comment (armijo1122) [reply test/test_inspect_utils.py]: Fixed！

**Review by aorenste — APPROVED** [review]

**Comment by aorenste** [comment]
Assuming tests pass, looks good to merge.

**Comment by armijo1122** [comment]
@aorenste I just fixed some lint error by using `linrunner -a`, but there's still some CI failure in `torch_openreg` build (CMake error) which seems unrelated to my Python changes in _fast_bind . It looks like an environment or infrastructure issue. I just rebased newest main branch. Could you please help re-trigger the CI, thanks!"

**Comment by armijo1122** [comment]
@zou3519 @aorenste I've fixed all the lint errors locally. I missed adding # noqa tags for the intentional mutable default arguments in the test case earlier. The rest of the CI checks have passed. Could you please take a look and approve when you have a chance? Thanks!

**Comment by armijo1122** [comment]
@laithsakka @soulitzer @aorenste Could you please help me re-trigger the CI? I would really appreciate it!

**Comment by armijo1122** [comment]
@laithsakka @aorenste Hi, sorry for the interruption! I am wondering how can I trigger CI without bothering you every time. Thank you! Do I need some permissions or anything else?
